ERCS/erc-7772.md

```diff
@@ -0,0 +1,271 @@
+
+``` solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+interface IGasSponsor {
+    event GasAuthorised(address indexed gasSponsor);
+    event GasAuthorisedError(address indexed gasSponsor, bytes memory reason);
+    event GasClaimed(address indexed gasSponsor, uint256 indexed gasClaimed);
+    event GasClaimedError(address indexed gasSponsor, bytes memory reason);
+
+    bytes32 constant SPONSOR_GAS_KEY = keccak256("sponsorGas.key");
+
+    function authoriseGas(bytes calldata data) external;
+    
+    function sponsorGas(uint256 amount) external;
+}
+```
+Interface: IGasSponsor
+- Purpose: To be implemented by dApps to manage gas funds internally.
+- Function: `sponsorGas`
+  - Role: Allows the Paymaster to claim gas funds from the dApp's contract.
+  - Parameters: 
+    - `uint256 amount`: The amount of gas to be sponsored.
+  - Visibility: External
+  - Usage: This function is called by the Paymaster during the post-operation (`postOp`) stage to claim the gas spent on the corresponding User Operation.
+- Function: `AuthoriseGas`
+  - Role: Allows the Paymaster to invoke this function in the dApp's contract during validation phase as an alternative pathway to approve the amount of gas willing to be sponsored. This function can be used when additional sponsors are not part of the execution phase but still wish to sponsor the transaction.
+  - Parameters: 
+    - `bytes calldata data`: Additional data that can be passed to the dApp to help the dApp determine the amount of gas it is willing to sponsor.
+  - Visibility: External
+  - Usage: This function is called by the Paymaster during the validatePaymasterUserOp stage to set the gas that the dApp is willing to sponsor for the User Operation.
+
+### DApp Example
+```solidity
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import "./interfaces/IGasSponsor.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+contract Dapp is IGasSponsor {
+
+    constructor(....) {
+        ....
+    }
+
+    function swap(....) public {
+        
+        ....
+        
+        assembly {
+            tstore(SPONSOR_GAS_KEY, 1000000000000000)
+        }
+    }
+    
+    function authoriseGas(
+        bytes calldata data
+    ) external override {
+        try {
+            // Any custom logic to decode the data
+            (
+                address someAddress,
+                bytes memory someData,
+                bytes memory someOtherData
+            ) = abi.decode(data, (address, bytes, bytes));
+            
+            if (...) {
+                assembly {
+                    tstore(SPONSOR_GAS_KEY, 1000000000000000)
+                }
+            }
+        } catch {
+            // Handle the case where data doesn't conform to the expected format
+            // Leave blank to not authorise any amount
+        }    
+    }
+
+    function sponsorGas(
```

Comment:
what prevent any contract in the batch to call sponsorGas ?
that is,  if a user make a complex batch call that goes through defiA then defiB (using some paymasterX) - and defiB simply calls defiA.sponsorGas
(and for that matter "defiB" can be any contract executed in the flow, like transfer target)